paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java

```diff
@@ -1363,4 +1395,104 @@ static ConflictCheck noConflictCheck() {
     static ConflictCheck mustConflictCheck() {
         return latestSnapshot -> true;
     }
+
+    private interface CommitResult {
+        boolean isSuccess();
+    }
+
+    private static class SuccessResult implements CommitResult {
+
+        @Override
+        public boolean isSuccess() {
+            return true;
+        }
+    }
+
+    private class RetryResult implements CommitResult {
+
+        private final String deltaManifestList;
+        private final String changelogManifestList;
+
+        private final String oldIndexManifest;
+        private final String newIndexManifest;
+
+        private final List<ManifestFileMeta> mergeBeforeManifests;
+        private final List<ManifestFileMeta> mergeAfterManifests;
```

Comment:
These are logically the same. What we really want to do is to optimize conflict checking.

Let's say we want to commit snapshot $s$, but the latest snapshot is snapshot $(s + k)$. Because we already confirmed that our new changes is compatible to snapshots up to $s$, during next retry we only need to check delta manifests in snapshots $(s + 1)$ to $(s + k)$.